examples/basic/eels_background/pipeline.json

```diff
@@ -54,24 +33,152 @@
+        },
+        "output": {}
+      },
+      "params": {
+        "filename": "spectrum.png",
+        "xLabel": "eLoss (eV)",
+        "yLabel": ""
+      }
+    },
+    {
+      "id": "21",
+      "image": "oremda/plot",
+      "queue": "plot",
+      "ports": {
+        "input": {
+          "x0": { "type": "data"},
+          "y0": { "type": "data"},
+          "x1": { "type": "data"},
+          "y1": { "type": "data"}
+        },
+        "output": {}
+      },
+      "params": {
+        "filename": "background.png",
+        "xLabel": "eLoss (eV)",
+        "yLabel": ""
+      }
+    },
+    {
+      "id": "22",
+      "image": "oremda/plot",
+      "queue": "plot",
+      "ports": {
+        "input": {
+          "x0": { "type": "data"},
+          "y0": { "type": "data"}
+        },
+        "output": {}
+      },
+      "params": {
+        "filename": "subtracted.png",
+        "xLabel": "eLoss (eV)",
+        "yLabel": ""
+      }
     }
   ],
   "edges": [
     {
+      "type": "data",
       "from": { "id": "10", "port": "eloss" },
-      "to": { "id": "0", "port": "eloss" }
+      "to": { "id": "20", "port": "x0" }
     },
     {
+      "type": "data",
       "from": { "id": "10", "port": "spec" },
-      "to": { "id": "0", "port": "spec" }
+      "to": { "id": "20", "port": "y0" }
     },
     {
-      "from": { "id": "0", "port": "eloss" },
+      "type": "data",
+      "from": { "id": "10", "port": "eloss" },
       "to": { "id": "11", "port": "eloss" }
     },
     {
-      "from": { "id": "0", "port": "spec" },
+      "type": "data",
+      "from": { "id": "10", "port": "spec" },
       "to": { "id": "11", "port": "spec" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "10", "port": "eloss" },
+      "to": { "id": "21", "port": "x0" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "10", "port": "spec" },
+      "to": { "id": "21", "port": "y0" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "11", "port": "eloss" },
+      "to": { "id": "21", "port": "x1" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "11", "port": "background" },
+      "to": { "id": "21", "port": "y1" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "10", "port": "eloss" },
+      "to": { "id": "12", "port": "eloss" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "10", "port": "spec" },
+      "to": { "id": "12", "port": "spec" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "11", "port": "eloss" },
+      "to": { "id": "12", "port": "eloss_bg" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "11", "port": "background" },
+      "to": { "id": "12", "port": "background" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "12", "port": "eloss" },
+      "to": { "id": "22", "port": "x0" }
+    },
+    {
+      "type": "data",
+      "from": { "id": "12", "port": "spec" },
+      "to": { "id": "22", "port": "y0" }
```

Comment:
Maybe we can reduce the number of times we have to specify `"type": "data"` by making it default to that if no type is given? That way, you don't have to specify a type if it is data.

I think having good defaults makes it easier for new users to get started (and less code for us to write). We might even be able to auto-generate some settings if they are not specified by inspecting other parts of the json (for the queue name, for instance).

Reply:
Good idea, we can default the type to `data` in general.

I think I'll simplify the json format quite a bike once I integrate the docker labels:
- Only the image name will be needed
- Port information and queue name can be inferred from the labels